### PR TITLE
fix(closure): patchOnProperty with exact eventNames as possible

### PR DESCRIPTION
**Pull request by JiaLiPassion**
related to #722.

We should try to patch onProperties with exact eventNames as possible.

in this PR, I separate the eventNames into several categories.

1. globalEventHandlerEventNames, eventNames in [GlobalEventHandlers](https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers)
2. documentEventNames, eventNames in [Document](https://developer.mozilla.org/en-US/docs/Web/API/Document)
3. window, eventNames in [Window](https://developer.mozilla.org/en-US/docs/Web/API/Window)
4. element, eventNames in [Element](https://developer.mozilla.org/en-US/docs/Web/API/Element)
5. webgl, eventNames in [WebGlContextEvent](https://developer.mozilla.org/en-US/docs/Web/API/WebGLContextEvent)
6. form, eventNames in [FormElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement)
7. details, eventNames in [DetailsElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDetailListElement)
8. XMLHttpRequest, eventNames in [XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest)
9. IDBIndex, eventNames in [IDBIndex](https://developer.mozilla.org/en-US/docs/Web/API/IDBIndex)
10. WebSocket, eventNames in [WebSocket](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket)

So we can patch each target with exact eventNames.

For example, we patch HTMLElement with 
`globalEventHandlers + element + webgl + form + details`, 
and we patch Window with
`globalEventHandlers + window`.

### CODE REVIEW

**Comment by JiaLiPassion** [comment]
@mhevery , just as #722, we may need to patch exact event names as possible instead of use `patchOnProperties(XXX.Prototype, null)`, please review whether the idea is OK or not, I will continue to check the event names.

File: lib/common/utils.ts [hunk 1]
```diff
@@ -148,10 +148,45 @@ export function patchProperty(obj: any, prop: string) {
...
+    }
+  }
+  return validEventNames;
+}
+
 export function patchOnProperties(obj: any, properties: string[]) {
   if (properties) {
-    for (let i = 0; i < properties.length; i++) {
-      patchProperty(obj, 'on' + properties[i]);
+    const validEventNames = getEventNames(obj, properties);
```
Comment (mhevery) [inline lib/common/utils.ts]: Could we have a fixed set rather than performing this complex operation on each startup?

Comment (JiaLiPassion) [inline lib/common/utils.ts]: @mhevery, yes, in my first commit, my goal is to make a fixed set, but it seems sometimes patch non exist property may cause errors, I will do more research and describe it more clear here.

File: lib/common/utils.ts [hunk 2]
```diff
@@ -60,6 +60,7 @@ export const isMix: boolean = typeof process !== 'undefined' &&
     !!(typeof window !== 'undefined' && (window as any)['HTMLElement']);
 
 export function patchProperty(obj: any, prop: string) {
+  // TODO: @JiaLiPassion, if the desc not exists in obj, should we still patch the property?
```
Comment (mhevery) [inline lib/common/utils.ts]: I don't think so. 

The way I imagine this to work is that we have a list of all properties, and than we only patch those properties, but only if the property already exist. But we don't look for extra ones.

Comment (JiaLiPassion) [inline lib/common/utils.ts]: @mhevery , got it, thanks! I also think we should do like that, I will modify it and do some tests.

File: lib/browser/property-descriptor.ts [hunk 3]
```diff
@@ -23,24 +230,32 @@ export function propertyDescriptorPatch(_global: any) {
   if (canPatchViaPropertyDescriptor()) {
     // for browsers that we can patch the descriptor:  Chrome & Firefox
     if (isBrowser) {
-      patchOnProperties(window, eventNames.concat(['resize']));
+      patchOnProperties(window, eventNames, Object.getPrototypeOf(window));
```
Comment (mhevery) [inline lib/browser/property-descriptor.ts]: The fact that we are getting all of the properties of the window, does not seem right. Why can't we have a fixed list here as well?

Comment (JiaLiPassion) [inline lib/browser/property-descriptor.ts]: @mhevery , the reason is in IE/Edge, the `onProp` don't exist in `window` but in `WindowPrototype`.

The code below describe the issue.

```javascript
const desc = Object.getOwnPropertyDescriptor(window, 'onload');
// in IE
console.log(desc); // will be undefined.
// in chrome/firefox
console.log(desc); // will be {configurable:true,get/set....}
```

So in new `patchProperty` method, we only patch the prop which exist in the target object, so I pass the prototype of `window` for the existency check.

**Pull request merged**